### PR TITLE
Added some new API functionality to Item List component

### PR DESCRIPTION
**Pull request by Krakean**
Added new signal: nothing_selected (when mouse clicked on empty space).
And new functions - unselect_all() & is_any_selected(), first one allows to deselect all currently selected items.

### CODE REVIEW

File: scene/gui/item_list.cpp [hunk 1]
```diff
@@ -28,8 +28,10 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 #include "item_list.h"
+#include "os/input.h"
 #include "os/os.h"
 #include "project_settings.h"
+#include <joypad.h>
```
Comment (akien-mga) [inline scene/gui/item_list.cpp]: Godot headers should be included with quotes. `< >` is for system headers.

**Comment by akien-mga** [comment]
Note: This is in preparation for a future PR to fix #13203 properly by allowing to deselect items in `FileDialog`, and thus handle an empty selection as current folder selection.

File: scene/gui/item_list.cpp (hunk 1 above)
Comment (Krakean) [reply scene/gui/item_list.cpp]: @akien-mga done

File: scene/gui/item_list.cpp [hunk 2]
```diff
@@ -28,6 +28,8 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 #include "item_list.h"
+#include "joypad.h"
```
Comment (akien-mga) [inline scene/gui/item_list.cpp]: Though I don't see why you need it? `"os/input.h"` should be enough.

Comment (Krakean) [reply scene/gui/item_list.cpp]: @akien-mga for backspace key constant.

Comment (akien-mga) [reply scene/gui/item_list.cpp]: That's defined in https://github.com/godotengine/godot/blob/30dadb12281924ccb59efc8df4b3b9ee68c6e3a4/core/os/keyboard.h not in joypad.h..

**Comment by vnen** [comment]
Why the `backspace_pressed` signal? I'm not fond of hardcoding the shortcut. At most you can set up a new `ui_` action, like `ui_back`. But I think this is better handled by specific instances of the control using `_gui_input()`.

**Comment by Krakean** [comment]
@vnen 
>Why the backspace_pressed signal?

I can rename it to proposed title.

>I'm not fond of hardcoding the shortcut. 

Agree, but I do not stop this shortcut anyhow or handle it in any specific way. It still passed "down there". Just subscribed control get notified that this key was pressed, and what they will do with it - depends on them. :) 
I do not nothing special in this regard (see tree.cpp->gui_input, and ask @reduz why he is doing the same with KEY_PAGEDOWN, KEY_UP and so on. In my case I just "catch" KEY_BACKSPACE in the same way)
Hhandle this particular case on editor_file_dialog side would be better choice, but I didn't found "central" space to catch backspace there, so decided to add new signal.

File: scene/gui/item_list.cpp (hunk 2 above)
Comment (Krakean) [reply scene/gui/item_list.cpp]: Well, Visual Studio suggested joypad.h :) Will fix it now.

**Comment by Noshyaar** [comment]
I don't think any Godot objects expose signal that refers to a specific keyboard button. Not even tree.cpp.

**Comment by Krakean** [comment]
@Noshyaar what if I rename this signal to something like "go back request"? :)

**Comment by vnen** [comment]
Well, common navigation keys (arrows, home, pgup, etc.) follow a certain standard and are expected to work everywhere. Backspace is just for deleting characters. For instance, browsers used to have backspace as a shortcut to go back, but they all abandoned it. (Even so, maybe Tree should be changed to use the `ui_` actions).

My major issue with this is that you are just emitting a signal when the key is pressed, the control itself isn't doing anything. For the control user, what's the difference between connecting to a signal and simply handling the input?

Also, ItemList is just a list of items, IMO it shouldn't have a concept of "go back". If you're doing this for the FileDialog, then those changes belong to the FileDialog control.

**Comment by Krakean** [comment]
@vnen okay then whats about other solution - what if I delete backspace_pressed signal, and replace it with general gui_input signal that will send received InputEvent in item_list's _gui_input()?

File: scene/gui/item_list.cpp [hunk 3]
```diff
@@ -28,6 +28,8 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 #include "item_list.h"
+#include "os/input.h"
+#include "os/keyboard.h"
```
Comment (akien-mga) [inline scene/gui/item_list.cpp]: Those can be removed as you removed the backspace_pressed signal.

Could you also amend the commit message (`git commit --amend`) to remove mentions of what is actually not in this commit anymore?

**Comment by Krakean** [comment]
@akien-mga Done. Should be fine now, check please.

**Pull request merged**